test/Streamly/Test/Data/Fold.hs

```diff
@@ -444,6 +450,133 @@ headAndRest ls = monadicIO $ do
     taill [] = []
     taill (_:xs) = xs
 
+partitionByM :: Property
+partitionByM =
+    forAll (listOf1 (chooseInt (intMin, intMax)))
+        $ \ls0 -> monadicIO $ action ls0
+
+    where
+
+    action ls = do
+        let f = \x -> if odd x then return (Left x) else return (Right x)
+        v1 <- run $ S.fold (F.partitionByM f FL.length FL.length) $ S.fromList ls
+        let v2 = foldl (\b a -> if odd a then b+1 else b) 0 ls
+            v3 = foldl (\b a -> if even a then b+1 else b) 0 ls
+        assert (v1 == (v2, v3))
+
+partitionBy :: Property
+partitionBy =
+    forAll (listOf1 (chooseInt (intMin, intMax)))
+        $ \ls0 -> monadicIO $ action ls0
+
+    where
+
+    action ls = do
+        let f = \x -> if odd x then Left x else Right x
+        v1 <- run
+                $ S.fold (F.partitionBy f FL.length FL.length)
+                $ S.fromList ls
+        let v2 = foldl (\b a -> if odd a then b+1 else b) 0 ls
+            v3 = foldl (\b a -> if even a then b+1 else b) 0 ls
+        assert (v1 == (v2, v3))
+
+demux :: Expectation
+demux =
+    let table = Data.Map.fromList [("SUM", F.sum), ("PRODUCT", F.product)]
+        input = Stream.fromList
+                [ ("SUM", 1::Int)
+                , ("PRODUCT", 2::Int)
+                , ("SUM",3)
+                , ("PRODUCT", 4::Int)
+                ]
+    in Stream.fold (F.demux table) input
+        `shouldReturn`
+        Data.Map.fromList [("PRODUCT", 8),("SUM", 4)]
+
+
+demuxWithSum :: Expectation
+demuxWithSum =
+    let f x = ("SUM", x::Int)
+        table = Data.Map.fromList [("SUM", F.sum)]
+        input = Stream.fromList [1, 4]
+
+    in Stream.fold (F.demuxWith f table) input
+        `shouldReturn`
+        Data.Map.fromList [("SUM", 5)]
+
+demuxWithProduct :: Expectation
+demuxWithProduct =
+    let f x = ("PRODUCT", x::Int)
+        table = Data.Map.fromList [("PRODUCT", F.product)]
+        input = Stream.fromList [2, 4]
+
+    in Stream.fold (F.demuxWith f table) input
+        `shouldReturn`
+        Data.Map.fromList [("PRODUCT", 8)]
+
+demuxDefaultWithSum :: Expectation
+demuxDefaultWithSum =
+    let f x = ("PRODUCT", x::Int)
+        table = Data.Map.fromList [("PRODUCT", FL.product)]
+        input = Stream.fromList [2, 4]
+
+    in Stream.fold (F.demuxDefaultWith f table (F.map snd FL.sum)) input
+        `shouldReturn`
+        (Data.Map.fromList [("PRODUCT" , 8)] , 0)
+
+demuxDefaultWithProduct :: Expectation
+demuxDefaultWithProduct =
+    let f x = ("PRODUCT", x::Int)
+        table = Data.Map.fromList [("PRODUCT", FL.product)]
+        input = Stream.fromList [2, 4]
+
+    in Stream.fold (F.demuxDefaultWith f table (F.map snd FL.product)) input
+        `shouldReturn`
+        (Data.Map.fromList [("PRODUCT" , 8)] , 1)
```

Comment:
For this, you should test for 2 cases,
1. When the table contains the default cases
2. When it does not
3. Basically test the default and the table

Reply:
Also, indentation.

Reply:
There is demuxWithProduct for this.